### PR TITLE
add kl divergence for forward_pass_logit_checker

### PR DESCRIPTION
**Pull request by ZhaoyueCheng**
## Description

+ add kl divergence option for forward_pass_logit_checker so that it compares the kl divergence of the output probability distribution with the golden output probability distribution
+ by default it's still using the original option to check numerical differences so it don't break other tests running

## Test
+ http://shortn/_XlXElFUgnv added the logs of 4 tests (kl passed, kl failed, numerical passed, numerical failed) with updated logging message here

### CODE REVIEW

File: MaxText/tests/forward_pass_logit_checker.py [hunk 1]
```diff
@@ -111,9 +111,22 @@ def main(config, test_args):
...
-    assert jax.numpy.allclose(
+
+    model_probabilities = jax.nn.softmax(full_train_logits[0, :token_size, :], axis=-1)
+    golden_probabilities = jax.nn.softmax(golden_logits[:token_size, :], axis=-1)
+
+    max_logging.log(f"{golden_probabilities[0]=}")
+    max_logging.log(f"{model_probabilities[0]=}")
+
+    kl_div = jax.numpy.sum(jax.scipy.special.kl_div(golden_probabilities, model_probabilities), axis=-1)
+    max_logging.log(f"{kl_div=}")
```
Comment (gagika) [inline MaxText/tests/forward_pass_logit_checker.py:122]: `max_logging.log(f"KL divergence={kl_div}") ` ?

**Comment by rdyro** [comment]
Great, thanks for implementing this! This looks good to me

File: MaxText/tests/forward_pass_logit_checker.py [hunk 2]
```diff
@@ -125,11 +138,12 @@ def main(config, test_args):
   parser.add_argument("--atol", type=float, required=False, default=0.1)
   parser.add_argument("--rtol", type=float, required=False, default=0.1)
   parser.add_argument("--token_size", type=int, required=False)
+  parser.add_argument("--kl", type=float, required=False, default=None)
```
Comment (gagika) [inline MaxText/tests/forward_pass_logit_checker.py:141]: Can you comment that if this is set then only KL divergence will be checked?

File: MaxText/tests/forward_pass_logit_checker.py [hunk 3]
```diff
@@ -111,9 +111,22 @@ def main(config, test_args):
...
+    golden_probabilities = jax.nn.softmax(golden_logits[:token_size, :], axis=-1)
+
+    max_logging.log(f"{golden_probabilities[0]=}")
+    max_logging.log(f"{model_probabilities[0]=}")
+
+    kl_div = jax.numpy.sum(jax.scipy.special.kl_div(golden_probabilities, model_probabilities), axis=-1)
+    max_logging.log(f"{kl_div=}")
+
+    if test_args.kl is not None:
+      assert jax.numpy.all(kl_div < test_args.kl)
```
Comment (gagika) [inline MaxText/tests/forward_pass_logit_checker.py:125]: Should we name it `max_kl_divergence` or `max_kl_div` and have a comment like:

f"KL Divergence ({kl_div}) exceeds the allowed threshold "
            f"({test_args.max_kl_divergence})"

**Review by gagika — APPROVED** [review]
thanks for the change, a few comments to improve logging, naming.

File: MaxText/tests/forward_pass_logit_checker.py (hunk 3 above)
Comment (gobbleturk) [reply MaxText/tests/forward_pass_logit_checker.py:125]: +1, clear error message very appreciated

**Review by gobbleturk — APPROVED** [review]

File: MaxText/tests/forward_pass_logit_checker.py (hunk 1 above)
Comment (ZhaoyueCheng) [reply MaxText/tests/forward_pass_logit_checker.py]: Thanks for the suggestion, updated!

File: MaxText/tests/forward_pass_logit_checker.py (hunk 3 above)
Comment (ZhaoyueCheng) [reply MaxText/tests/forward_pass_logit_checker.py]: Thanks for the suggestions, updated the logging 
http://shortn/_XlXElFUgnv added the logs of 4 sample tests here (kl passed, kl failed, numerical passed, numerical failed) with updated logging message

[the remark this answers is not in the document]
Comment (ZhaoyueCheng) [reply MaxText/tests/forward_pass_logit_checker.py]: Thanks for the suggestion, updated the logging here
http://shortn/_XlXElFUgnv added the logs of 4 sample tests here (kl passed, kl failed, numerical passed, numerical failed) with updated logging message

sample failed log:
```
KL divergence = [3.1939784e-05 2.4224353e-06 5.0701578e-06 5.3663748e-06] Max KL divergence = 3.1939784093992785e-05
Checking KL Divergence between train distribution and golden distribution
Traceback (most recent call last):
  File "/home/zhaoyuec/workdir/maxtext/MaxText/tests/forward_pass_logit_checker.py", line 154, in <module>
    main(cfg, test_args)
  File "/home/zhaoyuec/workdir/maxtext/MaxText/tests/forward_pass_logit_checker.py", line 126, in main
    assert jax.numpy.all(kl_div < test_args.max_kl_div), f"KL divergence values exceed the specified threshold of {test_args.max_kl_div}. Max divergence: {jax.numpy.max(kl_div)}"
AssertionError: KL divergence values exceed the specified threshold of 1e-07. Max divergence: 3.1939784093992785e-05
```

**Pull request merged**